File: src/id0077.c
// Licensed under the MIT License.

// Prime Summations

#include "../lib/euler.h"
#include "../lib/partition.h"
#include "../lib/sieve.h"
#define MAX_SEARCH 100

int main(void)
{
    long long term = -1;
    struct Sieve primes;
    clock_t start = clock();

    euler_ok(sieve(&primes, MAX_SEARCH));

    long long* constraints = primes.primes.items;
    long long* partitions = restricted_partition_range(
        MAX_SEARCH, 
        constraints,
        primes.primes.count);

    euler_assert(partitions);

    for (int n = 0; n < MAX_SEARCH; n++)
    {
        if (partitions[n] > 5000)
        {
            term = n;

            break;
        }
    }

    free(partitions);

    return euler_submit(77, term, start);
}

File: src/id0077.py
# Licensed under the MIT License.

# Prime Summations

from sympy import primerange
from time import time

MAX_SEARCH = 100
term = -1
partitions = [0] * MAX_SEARCH
partitions[0] = 1
start = time()

for p in primerange(MAX_SEARCH):
    for j in range(p, MAX_SEARCH):
        partitions[j] += partitions[j - p]

for n in range(len(partitions)):
    if partitions[n] > 5000:
        term = n        
        break
    
print(f"0077{term:>64}    {time() - start:.6f}")
